From 93cdc01e67ef777837739923057fb526382c0991 Mon Sep 17 00:00:00 2001
From: Brett Brotherton <bbrotherton@google.com>
Date: Thu, 29 Jan 2026 13:47:59 -0700
Subject: [PATCH] Revert "Add std::from_range constructors to
 base::flat_tree<>."

This reverts commit 4bf17b6c8ace3c66d83962c784ec0f8589e6b9bc.
---
 base/containers/flat_set_unittest.cc  | 93 ++-------------------------
 base/containers/flat_tree.h           | 36 -----------
 base/containers/flat_tree_unittest.cc | 64 +-----------------
 3 files changed, 8 insertions(+), 185 deletions(-)

diff --git a/base/containers/flat_set_unittest.cc b/base/containers/flat_set_unittest.cc
index 2b0bf7ac88..c02773931a 100644
--- a/base/containers/flat_set_unittest.cc
+++ b/base/containers/flat_set_unittest.cc
@@ -4,11 +4,8 @@
 
 #include "base/containers/flat_set.h"
 
-#include <list>
 #include <string>
-#include <vector>
 
-#include "base/containers/adapters.h"
 #include "base/memory/ptr_util.h"
 #include "base/test/move_only_int.h"
 #include "testing/gmock/include/gmock/gmock.h"
@@ -54,96 +51,18 @@ TEST(FlatSet, IncompleteType) {
   A a;
 }
 
-TEST(FlatSet, IteratorConstructor) {
-  int input_vals[] = {1, 1, 1, 2, 2, 2, 3, 3, 3};
-
-  // Copy-from iterators.
-  flat_set<int> cont(std::begin(input_vals), std::end(input_vals));
-  EXPECT_THAT(cont, ElementsAre(1, 2, 3));
-
-  std::list<MoveOnlyInt> input_list;
-  input_list.emplace_back(1);
-  input_list.emplace_back(1);  // Duplicate.
-  input_list.emplace_back(2);
-  input_list.emplace_back(3);
-  input_list.emplace_back(3);  // Duplicate.
-  input_list.emplace_back(4);
-
-  // Move-from iterators.
-  flat_set<MoveOnlyInt> orig(std::make_move_iterator(input_list.begin()),
-                             std::make_move_iterator(input_list.end()));
-
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(1)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(2)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(3)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(4)));
-
-  // List contains original number of moved-from elements.
-  EXPECT_THAT(input_list, ElementsAre(0, 0, 0, 0, 0, 0));
-}
-
 TEST(FlatSet, RangeConstructor) {
-  std::array<int, 9> input_vals = {1, 1, 1, 2, 2, 2, 3, 3, 3};
-
-  // Copy-from range.
-  flat_set<int> cont(std::from_range, input_vals);
-  EXPECT_THAT(cont, ElementsAre(1, 2, 3));
-
-  std::list<MoveOnlyInt> input_list;
-  input_list.emplace_back(1);
-  input_list.emplace_back(1);  // Duplicate.
-  input_list.emplace_back(2);
-  input_list.emplace_back(3);
-  input_list.emplace_back(3);  // Duplicate.
-  input_list.emplace_back(4);
-
-  // Move-from range
-  flat_set<MoveOnlyInt> orig(std::from_range,
-                             base::RangeAsRvalues(std::move(input_list)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(1)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(2)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(3)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(4)));
-
-  // List contains original number of moved-from elements.
-  EXPECT_THAT(input_list, ElementsAre(0, 0, 0, 0, 0, 0));
-}
-
-TEST(FlatSet, SortedUninqueRangeConstructor) {
-  std::array<int, 3> input_vals = {1, 2, 3};
+  flat_set<int>::value_type input_vals[] = {1, 1, 1, 2, 2, 2, 3, 3, 3};
 
-  // Copy-from range.
-  flat_set<int> cont(std::from_range, base::sorted_unique, input_vals);
+  flat_set<int> cont(std::begin(input_vals), std::end(input_vals));
   EXPECT_THAT(cont, ElementsAre(1, 2, 3));
-
-  std::list<MoveOnlyInt> input_list;
-  input_list.emplace_back(1);
-  input_list.emplace_back(2);
-  input_list.emplace_back(3);
-  input_list.emplace_back(4);
-
-  // Move-from list
-  flat_set<MoveOnlyInt> orig(std::from_range, base::sorted_unique,
-                             base::RangeAsRvalues(std::move(input_list)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(1)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(2)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(3)));
-  EXPECT_EQ(1U, orig.count(MoveOnlyInt(4)));
-
-  // List contains original number of moved-from elements.
-  EXPECT_THAT(input_list, ElementsAre(0, 0, 0, 0));
 }
 
 TEST(FlatSet, MoveConstructor) {
-  std::vector<MoveOnlyInt> input_vec;
-  input_vec.emplace_back(1);
-  input_vec.emplace_back(1);  // Duplicate.
-  input_vec.emplace_back(2);
-  input_vec.emplace_back(3);
-  input_vec.emplace_back(3);  // Duplicate.
-  input_vec.emplace_back(4);
-
-  flat_set<MoveOnlyInt> original(std::move(input_vec));
+  int input_range[] = {1, 2, 3, 4};
+
+  flat_set<MoveOnlyInt> original(std::begin(input_range),
+                                 std::end(input_range));
   flat_set<MoveOnlyInt> moved(std::move(original));
 
   EXPECT_EQ(1U, moved.count(MoveOnlyInt(1)));
diff --git a/base/containers/flat_tree.h b/base/containers/flat_tree.h
index 5104d90b4f..8fc8e78db4 100644
--- a/base/containers/flat_tree.h
+++ b/base/containers/flat_tree.h
@@ -169,12 +169,6 @@ class flat_tree {
             InputIterator last,
             const key_compare& comp = key_compare());
 
-  template <class Range>
-    requires(std::ranges::input_range<Range>)
-  flat_tree(std::from_range_t,
-            Range&& range,
-            const key_compare& comp = key_compare());
-
   flat_tree(const container_type& items,
             const key_compare& comp = key_compare());
 
@@ -190,13 +184,6 @@ class flat_tree {
             InputIterator last,
             const key_compare& comp = key_compare());
 
-  template <class Range>
-    requires(std::ranges::input_range<Range>)
-  flat_tree(std::from_range_t,
-            sorted_unique_t,
-            Range&& range,
-            const key_compare& comp = key_compare());
-
   flat_tree(sorted_unique_t,
             const container_type& items,
             const key_compare& comp = key_compare());
@@ -567,17 +554,6 @@ flat_tree<Key, GetKeyFromValue, KeyCompare, Container>::flat_tree(
   sort_and_unique();
 }
 
-template <class Key, class GetKeyFromValue, class KeyCompare, class Container>
-template <class Range>
-  requires(std::ranges::input_range<Range>)
-flat_tree<Key, GetKeyFromValue, KeyCompare, Container>::flat_tree(
-    std::from_range_t,
-    Range&& range,
-    const KeyCompare& comp)
-    : comp_(comp), body_(std::from_range, std::forward<Range>(range)) {
-  sort_and_unique();
-}
-
 template <class Key, class GetKeyFromValue, class KeyCompare, class Container>
 flat_tree<Key, GetKeyFromValue, KeyCompare, Container>::flat_tree(
     const container_type& items,
@@ -612,18 +588,6 @@ flat_tree<Key, GetKeyFromValue, KeyCompare, Container>::flat_tree(
   DCHECK(is_sorted_and_unique(*this, value_comp()));
 }
 
-template <class Key, class GetKeyFromValue, class KeyCompare, class Container>
-template <class Range>
-  requires(std::ranges::input_range<Range>)
-flat_tree<Key, GetKeyFromValue, KeyCompare, Container>::flat_tree(
-    std::from_range_t,
-    sorted_unique_t,
-    Range&& range,
-    const KeyCompare& comp)
-    : comp_(comp), body_(std::from_range, std::forward<Range>(range)) {
-  DCHECK(is_sorted_and_unique(*this, value_comp()));
-}
-
 template <class Key, class GetKeyFromValue, class KeyCompare, class Container>
 flat_tree<Key, GetKeyFromValue, KeyCompare, Container>::flat_tree(
     sorted_unique_t,
diff --git a/base/containers/flat_tree_unittest.cc b/base/containers/flat_tree_unittest.cc
index a2fe4cbfc3..be0e40b2c7 100644
--- a/base/containers/flat_tree_unittest.cc
+++ b/base/containers/flat_tree_unittest.cc
@@ -37,8 +37,6 @@
 #include <string>
 #include <vector>
 
-#include "base/compiler_specific.h"
-#include "base/containers/adapters.h"
 #include "base/containers/span.h"
 #include "base/test/gtest_util.h"
 #include "base/test/move_only_int.h"
@@ -332,7 +330,7 @@ TEST(FlatTree, MoveConstructor) {
 //           InputIterator last,
 //           const Compare& comp = Compare())
 
-TEST(FlatTree, IteratorConstructor) {
+TEST(FlatTree, RangeConstructor) {
   {
     IntPair input_vals[] = {{1, 1}, {1, 2}, {2, 1}, {2, 2}, {1, 3},
                             {2, 3}, {3, 1}, {3, 2}, {3, 3}};
@@ -355,30 +353,6 @@ TEST(FlatTree, IteratorConstructor) {
   }
 }
 
-// flat_tree(const Range& range,
-//           const Compare& comp = Compare())
-
-TEST(FlatTree, RangeConstructor) {
-  {
-    IntPair input_vals[] = {{1, 1}, {1, 2}, {2, 1}, {2, 2}, {1, 3},
-                            {2, 3}, {3, 1}, {3, 2}, {3, 3}};
-    auto span = base::span(input_vals);
-
-    IntPairTree first_of(std::from_range, span);
-    EXPECT_THAT(first_of,
-                ElementsAre(IntPair(1, 1), IntPair(2, 1), IntPair(3, 1)));
-  }
-  {
-    TreeWithStrangeCompare::value_type input_vals[] = {1, 1, 1, 2, 2,
-                                                       2, 3, 3, 3};
-    auto span = base::span(input_vals);
-
-    TreeWithStrangeCompare cont(std::from_range, span,
-                                NonDefaultConstructibleCompare(0));
-    EXPECT_THAT(cont, ElementsAre(1, 2, 3));
-  }
-}
-
 // flat_tree(const container_type&)
 
 TYPED_TEST_P(FlatTreeTest, ContainerCopyConstructor) {
@@ -445,7 +419,7 @@ TYPED_TEST_P(FlatTreeTest, InitializerListConstructor) {
 //           InputIterator last,
 //           const Compare& comp = Compare())
 
-TEST(FlatTree, SortedUniqueIteratorConstructor) {
+TEST(FlatTree, SortedUniqueRangeConstructor) {
   {
     IntPair input_vals[] = {{1, 1}, {2, 1}, {3, 1}};
     auto span = base::span(input_vals);
@@ -466,30 +440,6 @@ TEST(FlatTree, SortedUniqueIteratorConstructor) {
   }
 }
 
-// flat_tree(std::from_range_t,
-//           sorted_unique_t,
-//           const Range& range,
-//           const Compare& comp = Compare())
-
-TEST(FlatTree, SortedUniqueRangeConstructor) {
-  {
-    IntPair input_vals[] = {{1, 1}, {2, 1}, {3, 1}};
-    auto span = base::span(input_vals);
-
-    IntPairTree first_of(std::from_range, sorted_unique, span);
-    EXPECT_THAT(first_of,
-                ElementsAre(IntPair(1, 1), IntPair(2, 1), IntPair(3, 1)));
-  }
-  {
-    TreeWithStrangeCompare::value_type input_vals[] = {1, 2, 3};
-    auto span = base::span(input_vals);
-
-    TreeWithStrangeCompare cont(std::from_range, sorted_unique, span,
-                                NonDefaultConstructibleCompare(0));
-    EXPECT_THAT(cont, ElementsAre(1, 2, 3));
-  }
-}
-
 // flat_tree(sorted_unique_t, const container_type&)
 
 TYPED_TEST_P(FlatTreeTest, SortedUniqueContainerCopyConstructor) {
@@ -964,16 +914,6 @@ TEST(FlatTree, InsertRange) {
                                   IntPair(4, 1), IntPair(5, 2), IntPair(6, 2),
                                   IntPair(7, 2), IntPair(8, 2)));
   }
-  {
-    std::list<MoveOnlyInt> list;
-    list.emplace_back(1);
-    list.emplace_back(2);
-
-    MoveOnlyTree tree;
-    tree.insert_range(base::RangeAsRvalues(std::move(list)));
-    EXPECT_EQ(1u, tree.count(MoveOnlyInt(1)));
-    EXPECT_EQ(1u, tree.count(MoveOnlyInt(2)));
-  }
 }
 
 // template <class... Args>
-- 
2.53.0.rc1.225.gd81095ad13-goog

